packages/frontend/editor-ui/src/features/ai/assistant/builder.store.ts

```diff
@@ -112,6 +131,91 @@ export const useBuilderStore = defineStore(STORES.BUILDER, () => {
 		return creditsRemaining.value !== undefined ? creditsRemaining.value === 0 : false;
 	});
 
+	function extractPlaceholderLabel(value: unknown): string | null {
+		if (typeof value !== 'string') return null;
+		if (!value.startsWith(PLACEHOLDER_PREFIX) || !value.endsWith(PLACEHOLDER_SUFFIX)) return null;
+
+		const label = value
+			.slice(PLACEHOLDER_PREFIX.length, value.length - PLACEHOLDER_SUFFIX.length)
+			.trim();
+		return label.length > 0 ? label : null;
+	}
+
+	function findPlaceholderDetails(value: unknown, path: string[] = []): PlaceholderDetail[] {
+		const label = extractPlaceholderLabel(value);
+		if (label) return [{ path, label }];
+
+		if (Array.isArray(value)) {
+			return value.flatMap((item, index) => findPlaceholderDetails(item, [...path, `[${index}]`]));
+		}
+
+		if (value !== null && typeof value === 'object') {
+			return Object.entries(value).flatMap(([key, nested]) =>
+				findPlaceholderDetails(nested, [...path, key]),
+			);
+		}
+
+		return [];
+	}
```

Comment:
The size of this store is balooning, IMO these should be externalized to some composable